backend/opentelemetry.go

```diff
@@ -41,98 +42,42 @@ type OpenTelemetryBackend struct {
-	}
-
-	// Set up resource with service information
-	serviceName := "buildkite-agent-metrics"
-	if cfg.ServiceName != "" {
-		serviceName = cfg.ServiceName
-	}
-
-	serviceNamespace := "buildkite-agent-metrics"
-	if cfg.ServiceNamespace != "" {
-		serviceNamespace = cfg.ServiceNamespace
-	}
-
-	resourceAttrs := []attribute.KeyValue{
-		semconv.ServiceName(serviceName),
-		semconv.ServiceNamespace(serviceNamespace),
-	}
-	if cfg.ServiceVersion != "" {
-		resourceAttrs = append(resourceAttrs, semconv.ServiceVersion(cfg.ServiceVersion))
+// NewOpenTelemetryBackend creates a new OpenTelemetry backend.
+//
+// # Configuration is through standard OpenTelemetry SDK environment variables
+//
+// https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/
+// https://opentelemetry.io/docs/specs/otel/protocol/exporter/#endpoint-urls-for-otlphttp
+func NewOpenTelemetryBackend() (*OpenTelemetryBackend, error) {
+	serviceName := os.Getenv("OTEL_SERVICE_NAME")
+	if serviceName == "" {
+		serviceName = "buildkite-agent-metrics"
 	}
 
 	res := resource.NewWithAttributes(
 		semconv.SchemaURL,
-		resourceAttrs...,
+		semconv.ServiceName(serviceName),
+		semconv.ServiceNamespace("buildkite"),
+		semconv.ServiceVersion(version.Version),
 	)
 
 	// Set up trace exporter based on protocol
 	var traceExporter sdktrace.SpanExporter
 	var err error
-	
-	// For HTTP exporters, extract hostname from full URL if needed
-	httpEndpoint := cfg.Endpoint
-	if strings.HasPrefix(cfg.Endpoint, "https://") {
-		httpEndpoint = strings.TrimPrefix(cfg.Endpoint, "https://")
-	} else if strings.HasPrefix(cfg.Endpoint, "http://") {
-		httpEndpoint = strings.TrimPrefix(cfg.Endpoint, "http://")
+	protocol := os.Getenv("OTEL_EXPORTER_OTLP_PROTOCOL")
+	// default to http/protobuf as per spec
+	if protocol == "" {
+		protocol = "http/protobuf"
 	}
-	
-	if cfg.Protocol == "grpc" {
-		traceExporterOpts := []otlptracegrpc.Option{
-			otlptracegrpc.WithEndpoint(cfg.Endpoint),
-		}
-		if cfg.APIKey != "" {
-			traceExporterOpts = append(traceExporterOpts, otlptracegrpc.WithHeaders(map[string]string{
-				"authorization": cfg.APIKey,
-			}))
-		}
-		traceExporter, err = otlptracegrpc.New(context.Background(), traceExporterOpts...)
-	} else {
-		traceExporterOpts := []otlptracehttp.Option{
-			otlptracehttp.WithEndpoint(httpEndpoint),
-		}
-		// Only add URL path if endpoint doesn't already contain it
-		if !strings.Contains(httpEndpoint, "/v1/traces") {
-			traceExporterOpts = append(traceExporterOpts, otlptracehttp.WithURLPath("/v1/traces"))
-		}
-		if cfg.APIKey != "" {
-			traceExporterOpts = append(traceExporterOpts, otlptracehttp.WithHeaders(map[string]string{
-				"authorization": cfg.APIKey,
-			}))
-		}
-		traceExporter, err = otlptracehttp.New(context.Background(), traceExporterOpts...)
+
+	switch protocol {
+	case "grpc":
+		traceExporter, err = otlptracegrpc.New(context.Background())
```

Comment:
I think these `context.Background`s ought to be passed in as a `ctx` argument to this `NewOpenTelemetryBackend` func. Happy to note it as a todo (here, elsewhere, or even as a `context.TODO`).

Reply:
Great call

In e332fcf I added a `context.TODO` to the top of the to functions. To thread it through from the outside is a bit more involved as it changes the `backend.Backend` interface, but that would be a good followup.